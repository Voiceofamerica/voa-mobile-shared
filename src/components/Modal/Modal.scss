
@import "../../commonStyles/index";

$transition-time: 0.3s;

.modalContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  pointer-events: none;
  opacity: 0;
  transition: $transition-time opacity;

  &.show {
    pointer-events: auto;
    opacity: 1;

    .modal {
      transform: translateY(0);
      box-shadow: 0 3px 5px rgba(0, 0, 0, 0.5);
    }
  }

  .modal {
    width: 75vw;
    border-radius: 5px;
    overflow: hidden;
    z-index: $menu-z-max + 2;
    transition: $transition-time transform, $transition-time box-shadow;
    transform: translateY(20px);
    box-shadow: 0 0 rgba(0, 0, 0, 0.5);

    .content {
      padding: 10px;
    }

    .buttonList {
      display: flex;
      flex-direction: row;

      .button {
        flex: 1;
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: center;
        padding: 5px;
        font-size: 1.1em;
        font-weight: bold;
        border: none;
        border-top: 1px solid;
        cursor: pointer;
        user-select: none;

        &:not(:first-of-type) {
          border-left: 1px solid;
        }
      }
    }
  }

  .overlay {
    z-index: $menu-z-max + 1;
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
  }
}