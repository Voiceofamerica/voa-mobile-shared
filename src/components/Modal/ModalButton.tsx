
import * as React from 'react'

import { themed, ThemeProps, DEFAULT_THEME } from '../ThemeProvider'

import {
  button,
} from './Modal.scss'

export interface Props<T = {}> extends ThemeProps {
  id: T
  onClick?: (id: T) => void
  className?: string
  style?: React.CSSProperties
}

@themed
class ModalButton<T = {}> extends React.Component<Props<T>> {
  render () {
    const { id, onClick = () => null, className = '', style = {}, theme = DEFAULT_THEME, children } = this.props
    const { modalButtonBorder, modalButtonColor } = theme
    return (
      <div
        onClick={() => onClick(id)}
        className={`${button} ${className}`}
        style={{ borderColor: modalButtonBorder, color: modalButtonColor, ...style }}
      >
        {children}
      </div>
    )
  }
}

export default ModalButton
