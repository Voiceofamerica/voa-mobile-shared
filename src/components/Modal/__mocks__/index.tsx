
import * as React from 'react'

export default (props) => (
  React.createElement('modal', { props }, props.children)
)
export const ModalButton = (props) => (
  React.createElement('modal-button', { props }, props.children)
)
