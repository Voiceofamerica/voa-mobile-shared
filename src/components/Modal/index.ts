
export { default } from './Modal'
export { default as ModalButton } from './ModalButton'
