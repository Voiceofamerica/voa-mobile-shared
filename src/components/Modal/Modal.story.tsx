
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import ThemeProvider, { BaseTheme, ModalTheme } from '../ThemeProvider'

import Modal, { Props as ModalProps } from './Modal'
import ModalButton from './ModalButton'

class ModalDisplayer extends React.Component<ModalProps> {
  private modal: Modal | null

  render () {
    const { children, ...otherProps } = this.props

    return (
      <div>
        <Modal ref={this.setModal} {...otherProps}>{children}</Modal>
        <button type='button' onClick={this.showModal}>Show modal</button>
      </div>
    )
  }

  private showModal = () => {
    if (this.modal) {
      action('show modal')()
      this.modal.show().then(action('modal response'))
    }
  }

  private setModal = (el: Modal | null) => {
    this.modal = el
  }
}

const testBaseTheme: BaseTheme = {
  lightGrey: '#666666',
  black: '#DDDDFF',
  grey: '#AAAAAA',
  accentBlue: '#75A6D3',
}

const testLocalTheme: ModalTheme = {
  modalBackground: '#003300',
  modalColor: '#DDFFDD',
  modalButtonBorder: '#664499',
  modalButtonColor: '#CCCCFF',
  modalBackdropColor: '#152540',
}

storiesOf(Modal.name, module)
  .add('default display', () => (
    <ModalDisplayer>
      This is some modal content
    </ModalDisplayer>
  ))
  .add('default display clickable overlay', () => (
    <ModalDisplayer overlayClickable>
      Try clicking the backdrop
    </ModalDisplayer>
  ))
  .add('display with custom buttons', () => (
    <ModalDisplayer>
      This is some modal content
      <ModalButton id='one'>One</ModalButton>
      <ModalButton id={2}>Two</ModalButton>
      <ModalButton id={[3, 3, 3]}>Three</ModalButton>
    </ModalDisplayer>
  ))
  .add('base themed display', () => (
    <ThemeProvider value={testBaseTheme}>
      <ModalDisplayer>
        This is some modal content
      </ModalDisplayer>
    </ThemeProvider>
  ))
  .add('local themed display', () => (
    <ThemeProvider value={testLocalTheme}>
      <ModalDisplayer>
        This is some modal content
      </ModalDisplayer>
    </ThemeProvider>
  ))
