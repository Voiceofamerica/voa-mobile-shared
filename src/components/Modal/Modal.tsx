
import * as React from 'react'

import { toRGBAstring } from '../../helpers/colorHelper'

import { ThemeConsumer } from '../ThemeProvider'

import ModalButton, { Props as ModalButtonProps } from './ModalButton'
import {
  modalContainer,
  show,
  overlay,
  modal,
  content,
  buttonList,
} from './Modal.scss'

export interface Props {
  overlayClickable?: boolean
  className?: string
  style?: React.CSSProperties
}

export interface State {
  isOpen: boolean
}

export const CLOSE = 'close'

class Modal extends React.Component<Props, State> {
  state: State = {
    isOpen: false,
  }

  private openResolver?: (id: any) => void
  private openRejecter?: (id: any) => void

  show = (): Promise<any> => {
    return new Promise<any>((resolve, reject) => {
      this.openResolver = resolve
      this.openRejecter = reject
      this.toggleOpen(true)
    }).then((id) => {
      this.openResolver = undefined
      this.openRejecter = undefined
      this.toggleOpen(false)
      return id
    }, (err) => {
      this.openResolver = undefined
      this.openRejecter = undefined
      this.toggleOpen(false)
      return Promise.reject(err)
    })
  }

  render () {
    const { overlayClickable = false, className = '', style = {}, children } = this.props
    const { isOpen } = this.state
    console.log('renering modal', isOpen)

    const rChildren = React.Children.toArray(children)
    const buttons = (rChildren as React.ReactElement<ModalButtonProps>[]).filter(c => c.type === ModalButton)
    const modalContent = (rChildren as React.ReactElement<any>[]).filter(c => c.type !== ModalButton)

    const actualButtons = buttons.length > 0
                        ? this.mapButtons(buttons)
                        : [<ModalButton key={CLOSE} id={CLOSE} onClick={this.mapClick()}>Close</ModalButton>]

    return (
      <ThemeConsumer>
        {
          ({
            modalBackground: background,
            modalColor: color,
            modalBackdropColor,
          }) => (
            <div className={isOpen ? `${modalContainer} ${show}` : modalContainer} style={style}>
              <div
                className={overlay}
                onClick={overlayClickable ? () => this.mapClick()(CLOSE) : undefined}
                style={{ background: toRGBAstring(modalBackdropColor, 0.3) }}
              />
              <div className={`${modal} ${className}`} style={{ background, color }}>
                <div className={content}>
                  {modalContent}
                </div>
                <div className={buttonList}>
                  {actualButtons}
                </div>
              </div>
            </div>
          )
        }
      </ThemeConsumer>
    )
  }

  private toggleOpen = (isOpen: boolean = !this.state.isOpen) =>
    this.setState({ isOpen })

  private mapButtons = (buttons: React.ReactElement<ModalButtonProps>[]) =>
    buttons.map((button) => React.cloneElement(button, {
      ...button.props,
      key: button.props.id,
      onClick: this.mapClick(button.props.onClick),
    } as ModalButtonProps))

  private mapClick = (originalHandler: (id: any) => void = () => null) => (id: any) => {
    try {
      originalHandler(id)
      this.openResolver && this.openResolver(id)
    } catch (err) {
      this.openRejecter && this.openRejecter(err)
    }
  }
}

export default Modal
