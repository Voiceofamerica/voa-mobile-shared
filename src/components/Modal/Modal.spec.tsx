
import * as React from 'react'
import { create } from 'react-test-renderer'
import { mount } from 'enzyme'

import ModalButton, { Props as ModalButtonProps } from './ModalButton'
import Modal, { CLOSE } from './Modal'
import { overlay, button } from './Modal.scss'

describe(`<${Modal.name} />`, () => {
  describe('shapshots', () => {
    it(`should render with a default button with id '${CLOSE}'`, () => {
      const element = create((
        <Modal>
          This should show up
        </Modal>
      ))

      expect(element).toMatchSnapshot()
    })

    it(`should render with the given buttons`, () => {
      const element = create((
        <Modal>
          This should show up
          <ModalButton id={1}>One</ModalButton>
          <ModalButton id={2}>Two</ModalButton>
        </Modal>
      ))

      expect(element).toMatchSnapshot()
    })
  })

  describe('show', () => {
    it(`should, when overlayClickable is set, return a promise which resolves to '${CLOSE}' when the overlay is pressed`, () => {
      const wrapper = mount((
        <Modal overlayClickable>
          This should show up
        </Modal>
      ))

      const instance = wrapper.instance() as Modal
      const showPromise = instance.show().then(id => {
        expect(id).toBe(CLOSE)
      })

      wrapper.find(`.${overlay}`).simulate('click')

      return showPromise
    })
    it(`should return a promise which resolves to '${CLOSE}' when the default button is pressed`, () => {
      const wrapper = mount((
        <Modal>
          This should show up
        </Modal>
      ))

      const instance = wrapper.instance() as Modal
      const showPromise = instance.show().then(id => {
        expect(id).toBe(CLOSE)
      })

      wrapper
        .find<ModalButtonProps>(ModalButton)
        .find(`.${button}`)
        .simulate('click')

      return showPromise
    })
    it(`should return a promise which resolves to the button's id when a button is pressed`, () => {
      const expectedId = 'test id'

      const wrapper = mount((
        <Modal>
          This should show up
          <ModalButton id='some other id'>One</ModalButton>
          <ModalButton id={expectedId}>Two</ModalButton>
        </Modal>
      ))

      const instance = wrapper.instance() as Modal
      const showPromise = instance.show().then(id => {
        expect(id).toBe(expectedId)
      })

      wrapper
        .find<ModalButtonProps>(ModalButton)
        .filterWhere(btn => btn.props().id === expectedId)
        .find(`.${button}`)
        .simulate('click')

      return showPromise
    })
    it(`should call the button's original onClick once when the button is pressed`, () => {
      const expectedId = 'test id'
      const onClick = jest.fn()

      const wrapper = mount((
        <Modal>
          This should show up
          <ModalButton id='some other id' onClick={onClick}>One</ModalButton>
          <ModalButton id={expectedId} onClick={onClick}>Two</ModalButton>
        </Modal>
      ))

      wrapper
        .find<ModalButtonProps>(ModalButton)
        .filterWhere(btn => btn.props().id === expectedId)
        .find(`.${button}`)
        .simulate('click')

      expect(onClick).toHaveBeenCalledTimes(1)
    })
    it(`should call the button's original onClick with the button's id when the button is pressed`, () => {
      const expectedId = 'test id'
      const onClick = jest.fn()

      const wrapper = mount((
        <Modal>
          This should show up
          <ModalButton id='some other id' onClick={onClick}>One</ModalButton>
          <ModalButton id={expectedId} onClick={onClick}>Two</ModalButton>
        </Modal>
      ))

      wrapper
        .find<ModalButtonProps>(ModalButton)
        .filterWhere(btn => btn.props().id === expectedId)
        .find(`.${button}`)
        .simulate('click')

      expect(onClick).toHaveBeenCalledWith(expectedId)
    })
  })
})
