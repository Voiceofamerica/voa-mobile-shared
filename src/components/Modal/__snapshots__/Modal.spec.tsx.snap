// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> shapshots should render with a default button with id 'close' 1`] = `
<div
  className="modalContainer"
  style={Object {}}
>
  <div
    className="overlay"
    onClick={undefined}
    style={
      Object {
        "background": "rgba(0, 0, 0, 0.3)",
      }
    }
  />
  <div
    className="modal "
    style={
      Object {
        "background": "#EEEEEE",
        "color": "#000000",
      }
    }
  >
    <div
      className="content"
    >
      This should show up
    </div>
    <div
      className="buttonList"
    >
      <div
        className="button "
        onClick={[Function]}
        style={
          Object {
            "borderColor": "#B4BCC2",
            "color": "#2083C6",
          }
        }
      >
        Close
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> shapshots should render with the given buttons 1`] = `
<div
  className="modalContainer"
  style={Object {}}
>
  <div
    className="overlay"
    onClick={undefined}
    style={
      Object {
        "background": "rgba(0, 0, 0, 0.3)",
      }
    }
  />
  <div
    className="modal "
    style={
      Object {
        "background": "#EEEEEE",
        "color": "#000000",
      }
    }
  >
    <div
      className="content"
    >
      This should show up
    </div>
    <div
      className="buttonList"
    >
      <div
        className="button "
        onClick={[Function]}
        style={
          Object {
            "borderColor": "#B4BCC2",
            "color": "#2083C6",
          }
        }
      >
        One
      </div>
      <div
        className="button "
        onClick={[Function]}
        style={
          Object {
            "borderColor": "#B4BCC2",
            "color": "#2083C6",
          }
        }
      >
        Two
      </div>
    </div>
  </div>
</div>
`;
